File: app/src/main/java/ru/deltadelete/lab10/adapters/CountryAdapter.java
package ru.deltadelete.lab10.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import androidx.annotation.DrawableRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import ru.deltadelete.lab10.R;
import ru.deltadelete.lab10.database.entities.Country;
import ru.deltadelete.lab10.databinding.CountryItemBinding;

public class CountryAdapter extends ArrayAdapter<Country> {

    private final List<Country> items;
    @LayoutRes
    private static final int layout = R.layout.country_item;
    private final LayoutInflater inflater = LayoutInflater.from(getContext());

    public CountryAdapter(Context context, List<Country> items) {
        super(context, layout, items);
        this.items = items;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        CountryItemBinding binding;
        if (convertView == null) {
            binding = CountryItemBinding.inflate(inflater, parent, false);
        } else {
            binding = CountryItemBinding.bind(convertView);
        }

        binding.setItem(getItem(position));

        return binding.getRoot();
    }

    @Nullable
    @Override
    public Country getItem(int position) {
        return items.get(position);
    }

    @Override
    public void add(@Nullable Country object) {
        if (object != null) {
            items.add(object);
        }
        super.add(object);
    }

    @Override
    public void remove(@Nullable Country object) {
        if (object != null) {
            items.remove(object);
        }
        super.remove(object);
    }
}

File: app/src/main/java/ru/deltadelete/lab10/adapters/CountryAdapter.kt
package ru.deltadelete.lab10.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import com.squareup.picasso.Picasso
import ru.deltadelete.lab10.R
import ru.deltadelete.lab10.database.entities.Country
import ru.deltadelete.lab10.databinding.CountryItemBinding

public class CountryAdapter(context: Context, private val items: MutableList<Country>) :
    ArrayAdapter<Country>(context, R.layout.country_item, items) {
    @LayoutRes
    private val layout: Int = R.layout.country_item
    private val inflater: LayoutInflater = LayoutInflater.from(context)

    @Override
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding: CountryItemBinding = if (convertView == null) {
            DataBindingUtil.inflate(inflater, R.layout.country_item, parent, false)
        } else {
            DataBindingUtil.bind<CountryItemBinding>(convertView)!!
        }
        // val view: View = convertView ?: inflater.inflate(layout, parent, false)

        val item: Country = getItem(position)
        binding.item = item

//        val text: TextView = view.findViewById(R.id.text_view_name);
//        val image: ImageView = view.findViewById(R.id.image_view_flag);

//        text.text = item.name
//        Picasso.get()
//                .load(item.flagUrl)
//                .error(R.drawable.baseline_error_outline_24)
//                .into(image)
//
//        return view
        return binding.rootCountryItem
    }

    override fun getItem(position: Int): Country {
        return items[position]
    }

    override fun add(item: Country?) {
        item?.let {
            super.add(it)
        }
    }

    override fun remove(item: Country?) {
        item?.let {
            items.remove(it)
            super.remove(it)
        }
    }
}